### PR TITLE
VACMS-16910: DUW Branch, Year, Month logic

### PR DESCRIPTION
**Pull request by chriskim2311**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Summary
Adds Branch, Year, and Month questions to DUW logic flow.

Question logic flow: I realized we only have conditional information on the results page based on answers selected throughout the form. With that I decided to check for the next question based on the current question answer, and previously answered questions and navigate accordingly. 

Homepage URL: `/discharge-upgrade-instructions/introduction/`

## Related issue(s)
https://github.com/department-of-veterans-affairs/va.gov-cms/issues/16910

## Testing done
Created unit tests and tested locally

## Screenshots
What_was_your_branch_of_service____PACT_Act___Veterans_Affairs
What_was_your_branch_of_service____How_to_Apply_for_A_Discharge_Upgrade___Veterans_Affairs

What_year_were_you_discharged_from_the_military____How_to_Apply_for_A_Discharge_Upgrade___Veterans_Affairs
What_year_were_you_discharged_from_the_military____How_to_Apply_for_A_Discharge_Upgrade___Veterans_Affairs

What_month_were_you_discharged____How_to_Apply_for_A_Discharge_Upgrade___Veterans_Affairs
What_month_were_you_discharged____How_to_Apply_for_A_Discharge_Upgrade___Veterans_Affairs

## What areas of the site does it impact?

*(Describe what parts of the site are impacted **if** code touched other areas)*

### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

**Comment by chriskim2311** [comment]
@laflannery @thejordanwood  This PR is ready for your review!
http://ac9f0484a44ca55b0120d613e59dc10b.review.vetsgov-internal/discharge-upgrade-instructions/introduction

**Comment by laflannery** [comment]
@chriskim2311 I'm getting a 502 error, can you take a look when you get a min?

**Comment by chriskim2311** [comment]
@randimays I have 3 commented out unit tests wondering if you can run one of them locally and see if you can figure out why they are not working properly.

**Comment by laflannery** [comment]
@chriskim2311 A few comments:
1. I already mentioned this in slack - can you add an `href` to the Get started link on the introduction page?
2. The 2 select box questions (year and month discharged) we need to format like we do the Zip code screen in Income limits so they would look like the screenshot and then we would be able to force focus to the H1 like we are doing on the radio button question:
Screenshot 2024-04-04 at 2 05 49 PM 1
3. The middle part of the `<title>` currently says "PACT Act", can you update this to say "Discharge Upgrade"?
4. We need to update the focus management when an error occurs (currently we have an existing backlog ticket to this in PACT ACT) but the current behavior is that when you press Continue without making any selections, the error message is announced but keyboard focus remains on the Continue button. What should happen is that when there's a validation error when a user presses Continue, focus moves back to the containing element of the field with the error with tabindex="-1". This will make it so that, your first tab stop will now be the input where you can correct the error.

File: src/applications/discharge-wizard/tests/v2/DischargeMonth.unit.spec.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+// import sinon from 'sinon';
+// import { RESPONSES } from '../../constants/question-data-map';
+// import { ROUTES } from '../../constants';
+
+// import DischargeMonth from '../../components/v2/questions/DischargeMonth';
+
+// const mockStoreStandard = {
+//   getState: () => ({
+//     dischargeUpgradeWizard: {
+//       duwform: {
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/DischargeMonth.unit.spec.js:14]: This should be updated wherever you use it - it's `duwForm`. These syntax errors cause that not helpful generic test failure :(

File: src/applications/discharge-wizard/tests/v2/DischargeMonth.unit.spec.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+//   viewedIntroPage: true,
+// };
+
+// const propsNoIntroPage = {
+//   formResponses: {
+//     SERVICE_BRANCH: RESPONSES.ARMY,
+//     DISCHARGE_YEAR: null,
+//     DISCHARGE_MONTH: null,
+//   },
+//   setServicePeriod: () => {},
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/DischargeMonth.unit.spec.js:66]: This should be `setDischargeMonth`

File: src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+// const pushStub = sinon.stub();
+
+// const propsStandard = {
+//   formResponses: {
+//     SERVICE_BRANCH: RESPONSES.ARMY,
+//     DISCHARGE_YEAR: null,
+//     DISCHARGE_MONTH: null,
+//   },
+//   setServicePeriod: () => {},
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js:53]: This should be `setDischargeYear`

File: src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+// import sinon from 'sinon';
+// import { RESPONSES } from '../../constants/question-data-map';
+// import { ROUTES } from '../../constants';
+
+// import DischargeYear from '../../components/v2/questions/DischargeYear';
+
+// const mockStoreStandard = {
+//   getState: () => ({
+//     dischargeUpgradeWizard: {
+//       duwform: {
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js:14]: Update to `duwForm`

File: src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+//     },
+//   }),
+//   subscribe: () => {},
+//   dispatch: () => {},
+// };
+
+// const mockStoreNoIntroPage = {
+//   getState: () => ({
+//     dischargeUpgradeWizard: {
+//       duwform: {
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/DischargeYear.unit.spec.js:31]: Update to `duwForm`

File: src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      setFormError={setFormError}
+      shortName={shortName}
+      testId="duw-discharge_year"
+      options={yearOptions}
+      valueSetter={setDischargeYear}
+    />
+  );
+};
+
+DischargeYear.propTypes = {
```
Comment (randimays) [inline src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx:76]: Update PropTypes here, same as DischargeMonth

File: src/applications/discharge-wizard/tests/v2/ServiceBranch.unit.spec.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+// import { expect } from 'chai';
+// import sinon from 'sinon';
+// import { ROUTES } from '../../constants';
+
+// import ServiceBranch from '../../components/v2/questions/ServiceBranch';
+
+// const mockStoreStandard = {
+//   getState: () => ({
+//     dischargeUpgradeWizard: {
+//       duwform: {
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/ServiceBranch.unit.spec.js:13]: Update to `duwForm`

File: src/applications/discharge-wizard/tests/v2/ServiceBranch.unit.spec.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+//     },
+//   }),
+//   subscribe: () => {},
+//   dispatch: () => {},
+// };
+
+// const mockStoreNoIntroPage = {
+//   getState: () => ({
+//     dischargeUpgradeWizard: {
+//       duwform: {
```
Comment (randimays) [inline src/applications/discharge-wizard/tests/v2/ServiceBranch.unit.spec.js:26]: Update to `duwForm`

**Review by randimays — CHANGES_REQUESTED** [review]
Some observations:

1. I missed this in your previous PR:

Screenshot 2024-04-04 at 1 40 10 PM

There's a typo in the word "Department" 4 lines from the bottom.

2. Anywhere that we're using `uswds="true"` or `uswds`, we can remove this. If the component has a v3 version, it's default now and we don't need to add the prop. Double-check that styling/functionality are the same though, please.

3. The spacing between the question and options here is quite large. I know we are using web components, but is it possible to tighten up the spacing here?

Screenshot 2024-04-04 at 1 52 03 PM

4. This may be fixed by the formatting adjustment Laura suggested to mirror Income Limits, but I'm noticing a big size discrepancy between the questions on mobile

Screenshot 2024-04-04 at 1 56 05 PM
Screenshot 2024-04-04 at 1 56 00 PM

5. This first page should have a back button

Screenshot 2024-04-04 at 1 58 55 PM

6. Make sure you address whatever `Github Actions / Linting` warnings come up, if possible. We can always ignore the "something something platform import is deprecated" ones, but the others are usually helpful.

File: src/applications/discharge-wizard/actions/index.js [hunk 9]
```diff
@@ -1,4 +1,11 @@
-import { DW_UPDATE_FIELD } from '../constants';
+import {
+  DW_UPDATE_FIELD,
```
[the remark this answers is not in the document]
Comment (chriskim2311) [reply src/applications/discharge-wizard/actions/index.js:2]: This is the previous v1 action. Added a comment to distinguish

File: src/applications/discharge-wizard/routes.jsx [hunk 10]
```diff
@@ -19,7 +22,11 @@ const envChildRoutes = environment.isProduction()
       { path: 'guidance', component: GuidancePage },
       { path: 'request-dd214', component: RequestDD214 },
       // new routes for DUW v2
+      { path: 'introduction1', component: HomePage },
```
[the remark this answers is not in the document]
Comment (chriskim2311) [reply src/applications/discharge-wizard/routes.jsx:25]: For some weird reason I get a loading spinner for /introduction. This was for testing but will remove

[the remark this answers is not in the document]
Comment (randimays) [reply src/applications/discharge-wizard/routes.jsx:25]: Do you have `yarn watch` or `yarn watch --env entry=discharge-wizard` running? For some reason, when I try to run it from the app folder, it never resolves 🤔 Not sure what causes that though.

[the remark this answers is not in the document]
Comment (chriskim2311) [reply src/applications/discharge-wizard/routes.jsx:25]: Using yarn watch. Not sure why either but the route works in staging and in previous review instances

File: src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+import PropTypes from 'prop-types';
+import { connect } from 'react-redux';
+import { range } from 'lodash';
+
+import {
+  QUESTION_MAP,
+  SHORT_NAME_MAP,
+} from '../../../constants/question-data-map';
+import { updateDischargeYear } from '../../../actions';
+import { pageSetup } from '../../../../pact-act/utilities/page-setup';
```
Comment (randimays) [inline src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx:11]: This import should be udpated

File: src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    />
+  );
+};
+
+DischargeYear.propTypes = {
+  formResponses: PropTypes.object.isRequired,
+  router: PropTypes.shape({
+    push: PropTypes.func,
+  }),
+  setDischargeYear: PropTypes.func.isRequired,
```
Comment (randimays) [inline src/applications/discharge-wizard/components/v2/questions/DischargeYear.jsx:81]: Check out the lint warning here 👀

File: src/applications/discharge-wizard/components/v2/questions/shared/Dropdown.jsx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+        onPrimaryClick={onContinueClick}
+        onSecondaryClick={onBackClick}
+        continue
+        uswds
+      />
+    </div>
+  );
+};
+
+Dropdown.propTypes = {
```
Comment (randimays) [inline src/applications/discharge-wizard/components/v2/questions/shared/Dropdown.jsx:103]: Several lint warnings here 👀

**Review by randimays — CHANGES_REQUESTED** [review]

**Comment by laflannery** [comment]
From my previous comment, 1-3 look good and 5 we discussed and we don't need because it's only 1 question/page, However, the focus management still needs some work:
- One the 2 select box fields, the `tabindex="-1" is staying on the component which means that if I need to go back to the field for any reason I can't, because it's need removed from the tab order. Here's [an example of a test form](https://staging.va.gov/mock-form-patterns/mailing-address) where this is being done that is a good example
- Also, the first radio button question we need to have this same focus behavior on, [The first step of the mock form](https://staging.va.gov/mock-form-patterns/relationship-to-veteran) has a radio group as an example

**Comment by thejordanwood** [comment]
@chriskim2311 I agree with all of the spacing notes that Randi gave here. 

@laflannery There's a year box in the [past income limits flow](https://www.va.gov/health-care/income-limits/year). Should the year/month discharge questions look like this too?

Screenshot 2024-04-05 at 11 16 10 AM

**Comment by laflannery** [comment]
@thejordanwood Yes, this is correct. Chris changed both of these on his latest updates to match what you showed  - the only difference is that we don't need the visual "Required" indicator since this is the only question on the screen (technically we don't need it on Income limits either)
Screenshot 2024-04-05 at 12 20 17 PM

**Comment by thejordanwood** [comment]
@chriskim2311 I see you just updated the screenshots–thank you!

1. I think the spacing between the header and the fields is still a bit too large. For income limits, it's about 32px of space so we should stay consistent with that. 
2. If we're staying consistent with income limits, the spacing between the fields and the back/continue buttons should be 16px. 
3. Looks like @laflannery answered my question about the year/month fields. No need to include the "Required" indicator. Can we change the field box to say "Select a year" or "Select a month" instead of "-Select-"?
4. About the error messages: What you have here is correct, but they all need periods at the end. Also we may need to say "Select a branch of service." instead of just branch. 

Apologies for the delayed review on my part!

**Comment by thejordanwood** [comment]
@chriskim2311 In that case, then we can move forward with the spacing as is.

**Review by randimays — APPROVED** [review]
From my earlier comments:

1. ✅ 
2. ✅ 
3. I realized this weird spacing is the same on PACT Act. Nothing we can do here ✅
4. ✅ 
5. ✅ 
6. ✅ 

LGTM, great work! 💥

**Pull request merged**